src/goto-programs/goto_convert.cpp

```diff
@@ -531,9 +531,7 @@ void goto_convertt::convert(
   // make sure dest is never empty
   if(dest.instructions.empty())
   {
-    dest.add_instruction(SKIP);
-    dest.instructions.back().code.make_nil();
-    dest.instructions.back().source_location=code.source_location();
+    dest.add(goto_programt::make_skip(code.source_location()));
```

Comment:
Looking at the use of these, couldn't we change the approach a bit and make `goto_programt::add` be a templated function taking a variadic number of arguments, with specialisations for the various instruction types?
````c++
template <goto_program_instruction_typet T, typename ... Params>
add(Params...);
template <> add<SKIP, source_locationt> (source_locationt loc) { ... };
````
My rationale is that we might eventually forbid to construct an `instructiont` other than via `.add`, and it also saves us typing `goto_programt::`. The above would then be `dest.add<SKIP>(code.source_location());`

Reply:
Yes, I've been tempted by that idea. Will explore.

Reply:
It should be ok to merge all the `make_X` work as that does the hard work of collecting individual assignments into a single function call. Possibly moving to `add<T>(...)` is then just text transformation.